### ISSUE TITLE
Failed to find exact match for etcd.database.coreos.com/v1beta2.EtcdCluster by [kind, name, singularName, shortNames]

### ISSUE THREAD

**Issue by schegi** [ansible/ansible#58001]
##### SUMMARY
Trying to deploy an Etcd cluster using the coreos etcd-operators custom resource EtcdCluster. 
Additionally a simple query for resources of this kind using k8s_facts also does not work, it returns an empty list. (resource previously deployed using kubectl apply.)

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

k8s
k8s_facts

##### ANSIBLE VERSION

```
ansible 2.8.1
  config file = None
  configured module search path = ['/Users/schegi/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/schegi/.virtualenvs/ansible/lib/python3.7/site-packages/ansible
  executable location = /Users/schegi/.virtualenvs/ansible/bin/ansible
  python version = 3.7.3 (default, Mar 27 2019, 09:23:15) [Clang 10.0.1 (clang-1001.0.46.3)]
```

##### CONFIGURATION

```paste below

```

##### OS / ENVIRONMENT
Darwin XXX.local 18.6.0 Darwin Kernel Version 18.6.0: Thu Apr 25 23:16:27 PDT 2019; root:xnu-4903.261.4~2/RELEASE_X86_64 x86_64

###### Required python packages
Name: kubernetes
Version: 8.0.1
Summary: Kubernetes python client
Home-page: https://github.com/kubernetes-client/python
Author: Kubernetes
Author-email: UNKNOWN
License: Apache License Version 2.0
Location: /Users/schegi/.virtualenvs/ansible/lib/python3.7/site-packages
Requires: setuptools, six, websocket-client, pyyaml, google-auth, certifi, urllib3, python-dateutil, requests, requests-oauthlib, adal
Required-by: openshift

Name: openshift
Version: 0.8.9
Summary: OpenShift python client
Home-page: https://github.com/openshift/openshift-restclient-python
Author: OpenShift
Author-email: UNKNOWN
License: Apache License Version 2.0
Location: /Users/schegi/.virtualenvs/ansible/lib/python3.7/site-packages
Requires: urllib3, dictdiffer, python-string-utils, six, kubernetes, jinja2, ruamel.yaml
Required-by:

###### Requirements:
Kubernetes cluster with coreos etcd-operator installed, e.g. via helm https://github.com/helm/charts/tree/master/stable/etcd-operator

Example values for etcd-operator helm chart:
```yaml
---
etcdOperator:
  commandArgs:
    cluster-wide: true
  image:
    tag: v0.9.4

backupOperator:
  image:
    tag: v0.9.4

restoreOperator:
  image:
    tag: v0.9.4
```

###### In cluster setup
Predeploy etcd cluster with
```bash
kubectl apply -f etcd-cluster.yml
```
etcd-cluster.yaml
```yaml
---
apiVersion: "etcd.database.coreos.com/v1beta2"
kind: "EtcdCluster"
metadata:
  name: "pre-deployed-etcd-cluster"
  namespace: "etcd"
  annotations:
    etcd.database.coreos.com/scope: clusterwide
spec:
  size: 3
  version: "3.3.13"
```

Example playbook:
```yaml
---
- name: Test install


  hosts: localhost
  gather_facts: false

  tasks:
    - name: test k8s_facts
      k8s_facts:
        api_version: "etcd.database.coreos.com/v1beta2"
        kind: "EtcdCluster"
      register: result

    - debug:
        var: result

    - name: test k8s etcdcluster resource
      k8s:
        definition:
          apiVersion: "etcd.database.coreos.com/v1beta2"
          kind: "EtcdCluster"
          metadata:
            name: "etcd-cluster"
            annotations:
              etcd.database.coreos.com/scope: clusterwide
          spec:
            size: 3
            version: "3.3.13"
```

##### EXPECTED RESULTS

Installed custom resource of kind etcdcluster is listed.
A custom resource of kind EtcdCluster is created. 

##### ACTUAL RESULTS

No resource of kind etcd cluster is listed and K8s module produces an error

```
TASK [test k8s_facts] **********************************************************************************************************************************************************************************************
task path: /Users/schegi/k8s_ops/ansible-setup/playbooks/test.yml:41
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: schegi
<127.0.0.1> EXEC /bin/sh -c 'echo ~schegi && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /Users/schegi/.ansible/tmp/ansible-tmp-1560868571.033664-189226410185626 `" && echo ansible-tmp-1560868571.033664-189226410185626="` echo /Users/schegi/.ansible/tmp/ansible-tmp-1560868571.033664-189226410185626 `" ) && sleep 0'
Using module file /Users/schegi/.virtualenvs/ansible/lib/python3.7/site-packages/ansible/modules/clustering/k8s/k8s_facts.py
<127.0.0.1> PUT /Users/schegi/.ansible/tmp/ansible-local-71097fmko98sd/tmpegeifxei TO /Users/schegi/.ansible/tmp/ansible-tmp-1560868571.033664-189226410185626/AnsiballZ_k8s_facts.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /Users/schegi/.ansible/tmp/ansible-tmp-1560868571.033664-189226410185626/ /Users/schegi/.ansible/tmp/ansible-tmp-1560868571.033664-189226410185626/AnsiballZ_k8s_facts.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/Users/schegi/.virtualenvs/ansible/bin/python3.7 /Users/schegi/.ansible/tmp/ansible-tmp-1560868571.033664-189226410185626/AnsiballZ_k8s_facts.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /Users/schegi/.ansible/tmp/ansible-tmp-1560868571.033664-189226410185626/ > /dev/null 2>&1 && sleep 0'
ok: [localhost] => {
    "changed": false,
    "invocation": {
        "module_args": {
            "api_key": null,
            "api_version": "etcd.database.coreos.com/v1beta2",
            "ca_cert": null,
            "client_cert": null,
            "client_key": null,
            "context": null,
            "field_selectors": [],
            "host": null,
            "kind": "EtcdCluster",
            "kubeconfig": null,
            "label_selectors": [],
            "name": null,
            "namespace": null,
            "password": null,
            "username": null,
            "validate_certs": null
        }
    },
    "resources": []
}

TASK [debug] *******************************************************************************************************************************************************************************************************
task path: /Users/schegi/k8s_ops/ansible-setup/playbooks/test.yml:47
ok: [localhost] => {
    "result": {
        "changed": false,
        "failed": false,
        "resources": []
    }
}

TASK [test k8s etcdcluster resource] *******************************************************************************************************************************************************************************
task path: /Users/schegi/k8s_ops/ansible-setup/playbooks/test.yml:50
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: schegi
<127.0.0.1> EXEC /bin/sh -c 'echo ~schegi && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /Users/schegi/.ansible/tmp/ansible-tmp-1560868574.8431518-194828762998139 `" && echo ansible-tmp-1560868574.8431518-194828762998139="` echo /Users/schegi/.ansible/tmp/ansible-tmp-1560868574.8431518-194828762998139 `" ) && sleep 0'
Using module file /Users/schegi/.virtualenvs/ansible/lib/python3.7/site-packages/ansible/modules/clustering/k8s/k8s.py
<127.0.0.1> PUT /Users/schegi/.ansible/tmp/ansible-local-71097fmko98sd/tmpcs2c1hc9 TO /Users/schegi/.ansible/tmp/ansible-tmp-1560868574.8431518-194828762998139/AnsiballZ_k8s.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /Users/schegi/.ansible/tmp/ansible-tmp-1560868574.8431518-194828762998139/ /Users/schegi/.ansible/tmp/ansible-tmp-1560868574.8431518-194828762998139/AnsiballZ_k8s.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/Users/schegi/.virtualenvs/ansible/bin/python3.7 /Users/schegi/.ansible/tmp/ansible-tmp-1560868574.8431518-194828762998139/AnsiballZ_k8s.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /Users/schegi/.ansible/tmp/ansible-tmp-1560868574.8431518-194828762998139/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
  File "/var/folders/10/bnmky9pj2gs8dtjl6bgr4l3m0000gn/T/ansible_k8s_payload_qhrzxiy2/ansible_k8s_payload.zip/ansible/module_utils/k8s/common.py", line 193, in find_resource
    return self.client.resources.get(api_version=api_version, short_names=[kind])
  File "/Users/schegi/.virtualenvs/ansible/lib/python3.7/site-packages/openshift/dynamic/client.py", line 641, in get
    raise ResourceNotFoundError('No matches found for {}'.format(kwargs))

fatal: [localhost]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "api_key": null,
            "api_version": "v1",
            "append_hash": false,
            "ca_cert": null,
            "client_cert": null,
            "client_key": null,
            "context": null,
            "definition": {
                "apiVersion": "etcd.database.coreos.com/v1beta2",
                "kind": "EtcdCluster",
                "metadata": {
                    "annotations": {
                        "etcd.database.coreos.com/scope": "clusterwide"
                    },
                    "name": "etcd-cluster"
                },
                "spec": {
                    "size": 3,
                    "version": "3.3.13"
                }
            },
            "force": false,
            "host": null,
            "kind": null,
            "kubeconfig": null,
            "merge_type": null,
            "name": null,
            "namespace": null,
            "password": null,
            "resource_definition": {
                "apiVersion": "etcd.database.coreos.com/v1beta2",
                "kind": "EtcdCluster",
                "metadata": {
                    "annotations": {
                        "etcd.database.coreos.com/scope": "clusterwide"
                    },
                    "name": "etcd-cluster"
                },
                "spec": {
                    "size": 3,
                    "version": "3.3.13"
                }
            },
            "src": null,
            "state": "present",
            "username": null,
            "validate": null,
            "validate_certs": null,
            "wait": false,
            "wait_condition": null,
            "wait_timeout": 120
        }
    },
    "msg": "Failed to find exact match for etcd.database.coreos.com/v1beta2.EtcdCluster by [kind, name, singularName, shortNames]"
}

PLAY RECAP *********************************************************************************************************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
```

**Comment by ansibot** [comment]
Files identified in the description:
* [`lib/ansible/modules/clustering/k8s/k8s.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/clustering/k8s/k8s.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @chouseknecht @fabianvf @flaper87 @maxamillion @willthames
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by fabianvf** [comment]
This is very strange, can you paste `kubectl get crds` and `ansible localhost -m 'k8s_facts' -a "api_version=apiextensions.k8s.io/v1beta1 kind=CustomResourceDefinition" -v`?

**Comment by schegi** [comment]
kubectl get crds:
etcdbackups.etcd.database.coreos.com             2019-05-14T13:28:29Z
etcdclusters.etcd.database.coreos.com            2019-05-14T13:28:29Z
etcdrestores.etcd.database.coreos.com            2019-05-14T13:28:29Z
volumesnapshotclasses.snapshot.storage.k8s.io    2019-05-27T08:08:00Z
volumesnapshotcontents.snapshot.storage.k8s.io   2019-05-27T08:08:00Z
volumesnapshots.snapshot.storage.k8s.io          2019-05-27T08:08:00Z

The otherb output is quite large, i'll go for a pastebin:
https://pastebin.com/ganSEMSX

**Comment by schegi** [comment]
Upgraded openshift to 0.9.0 and kubernetes to 9.0.0 and downgraded to serveral version down to 0.8.6/8.0.1 but always the same result.

**Comment by willthames** [comment]
That pastebin doesn't seem to include any reference to the string `etcd` at all (ignoring case).

Can you do `kubectl get crd etcdclusters -o yaml` (or maybe `etcdclusters.etcd.database.coreos.com`)

**Comment by schegi** [comment]
Seems like the error was partly on my side and/or k8s and k8s_facts are not capable of dealing with my kubeconfig setup. I have a KUBECONFIG env variable defining multiple kubeconfigs, kubectl can work with such a setup. Apparently k8s, k8s_facts cannot and always picking the first path. Therefore i close this issue here, thy for your time and sry should have taken a closer look at the problem cause.

**Comment by fabianvf** [comment]
@schegi ah I think we're currently picking the kubeconfig up from the `K8S_AUTH_KUBECONFIG` environment variable, we should probably be checking just `KUBECONFIG` as well. Feel free to open up a separate issue for that behavior.
